### PR TITLE
Bun.markdown: implement the latexMath and underline options

### PR DESCRIPTION
**Pull request by robobun**
### Problem

- `Bun.markdown`'s `latexMath` and `underline` options do nothing, although both are documented (docs/runtime/markdown.mdx, bun.d.ts) and the terminal preset used by `Bun.markdown.ansi()` and `bun ./file.md` lists both:
  - `Bun.markdown.html("$a+b$", { latexMath: true })` returns `<p>$a+b$</p>` (documented: `<x-equation>a+b</x-equation>`)
  - `Bun.markdown.html("_x_", { underline: true })` returns `<p><em>x</em></p>` (documented: `<u>x</u>`)
- `latexMath`: `src/md/parser.rs` adds `$` to the mark character map, but nothing in `src/md/inlines.rs` handles a `$` mark, so it falls through as text. Nothing ever emits `SpanType::Latexmath` / `LatexmathDisplay`.
- `underline`: `Options::to_flags()` in `src/md/root.rs` drops the field because `Flags` (`src/md/types.rs`) has no `underline`, and nothing ever emits `SpanType::U`.
- Neither option was ever implemented: the Zig parser this crate was ported from had the same gap, so this is not a regression. The HTML, ANSI, React and callback renderers all already handle the three span types; only the parser side is missing. md4c's test files for both extensions (`test/js/bun/md/spec-latex-math.txt`, `spec-underline.txt`) are checked in but were not in `md-spec.test.ts`'s list.

### Fix

- `src/md/inlines.rs`, `collect_emphasis_delimiters`: with `latex_math`, a run of one or two `$` becomes a delimiter entry (`emph_char == b'$'`). It can open if it starts the slice or follows whitespace/punctuation, and close if it ends the slice or precedes whitespace/punctuation (md4c's rule; `$5 and $10` stays text). Longer runs are text.
- `resolve_emphasis_delimiters` hands `$` entries to the new `resolve_math_delim`, which is md4c's `md_analyze_dollar`: a closer pairs with the most recent pending opener only if both runs have the same length (`$` vs `$$`), otherwise it becomes a pending opener itself; a match consumes both runs, deactivates every delimiter between them, records the closer position on the opener (`EmphDelim::close_pos`, the one new field) and discards all pending openers, so spans never nest. When an emphasis/strikethrough pair resolves, pending `$` openers inside it are discarded too (md4c's `md_pop_openers`), so emphasis and math cannot cross: whichever closes first wins, as in md4c.
- The emit walk renders a matched opener as one atomic span: `enter_span(Latexmath | LatexmathDisplay)`, the raw bytes up to the closer as `TextType::Latexmath`, `leave_span`, then skips past the closer. Unmatched runs are text. Because a span is only ever emitted at its opener, output is well formed even when another construct consumed the opener first.
- `src/md/html_renderer.rs` and `src/md/ansi_renderer.rs`: `TextType::Latexmath` shares the `TextType::Code` arm, and in both renderers a newline inside the span becomes a space (md4c's output for the multi-line example in spec-latex-math.txt). The ANSI renderer did not do this for code spans either: a span continuing on the next line was written verbatim, so the continuation lost its list indent or quote gutter and the column count went stale. That was already wrong for multi-line code spans on main and would have become reachable for math through the terminal preset; the shared arm fixes both.
- `underline`: `Flags` gains the field, `to_flags` passes it, and the emit walk gets an underline arm next to the strikethrough one: for a `_` run with the flag set it emits one `U` span per consumed underscore, straight from the run's `close_count` / `open_count`, instead of the em/strong tags. Pairing is unchanged, so intra-word underscores, the rule of three, etc. behave as before. `__x__` therefore renders `<u><u>x</u></u>`, which is md4c's documented output (spec-underline.txt example 2 has `___foo___` as three nested `<u>`). Using the counts rather than the per-run match-size arrays also keeps long runs exact: those arrays hold at most `MAX_EMPH_MATCHES` (6) entries, which is a pre-existing limitation of the em/strong output (#39496) that underline does not inherit.
- Terminal preset (`Options::TERMINAL`, used by `ansi()` and `bun ./file.md`): `latex_math` stays on, so math spans are now styled there (`$x$` in magenta). `underline` is switched off explicitly. It was listed in the preset but never worked, so terminal output has always rendered `_x_` / `__x__` as italic / bold; making it live would change that for every document that uses underscore emphasis, with no opt-out (`ansi()` takes no parser options), and the preset author's intent (md4c's all-underscores semantic, or only `__x__` as the old docs table said) is not recorded. #39560 holds that decision; it also notes the ANSI `U` styling bug that would need fixing first. Two tests pin the terminal behavior as unchanged (`ansi() keeps underscores as emphasis`, and the `bun ./file.md` case). The `ansi()` JSDoc no longer claims underline.
- Why this is the right behavior: the parser is a port of md4c, the renderers were written against md4c's span model (`<x-equation>` is md2html's output, `"math"` / `display: true` in the React renderer, `$` styling in the ANSI renderer), and md4c's own test files for both extensions are already in the tree. The implementation was checked against md4c's current `md_collect_marks` / `md_analyze_dollar` / `md_process_inlines`; both spec files pass unmodified.
- Cost when the options are off: `$` is not a mark character and `collect_emphasis_delimiters` never pushes `$` entries, so the default parsing path only sees `EmphDelim` grow from 64 to 72 bytes (it is copied once per block that contains emphasis delimiters); the ANSI renderer additionally scans each code span's bytes once for a newline. Every new resolution step is O(1) amortized; doubling the input doubles the time for six adversarial delimiter floods (see details).
- Docs and types: the options table and the React overrides table (`u`, `math`) in docs/runtime/markdown.mdx, the JSDoc for both options, and a `MathProps` type carrying the `display` prop the React renderer already sets. `test/integration/bun-types/fixture/markdown.ts` compiles realistic usage.
- Not changed here, tracked separately: `hardSoftBreaks` and `collapseWhitespace` are also documented options that do nothing (#39491, being implemented in #39495); emphasis delimiters inside a wiki link target can pair with delimiters outside it (#39492, pre-existing, fixed in #39494). This PR and #39494 are independent and land in either order: a math span is only ever emitted at its opener, so output stays well formed, and #39494's skip in the collection phase covers `$` runs too. Whichever of the two lands second should add exact cases for `$` / `_` runs on both sides of a wiki link boundary (noted on #39494). Em/strong output of a run with more than six matches is wrong on main independently of this PR (#39496). The terminal underline decision is #39560.
- Verification (the new tests fail on the unfixed build and pass with the fix; the negative-contract cases, such as options off and the preset's unchanged underscore rendering, pass both ways by design):
  - `test/js/bun/md/md-spec.test.ts` now runs `spec-latex-math.txt` and `spec-underline.txt` (6 examples fail before)
  - `test/js/bun/md/md-edge-cases.test.ts`: `latexMath option` / `underline option` blocks cover both flags on and off, both option spellings, precedence against code spans, links, images, emphasis, strikethrough, escapes, headings and table cells, non-ASCII neighbors, the `ansi()` preset, and a linear-time flood test, plus two long-run cases (14 underscores, and one run closed by seven closers) that the match-size arrays would truncate
  - `ansi(): line breaks inside math and code spans become spaces` (md-edge-cases.test.ts): list item, blockquote, soft-wrapped inline math, a multi-line code span (fails on main today) and a `columns` case; `ansi() keeps underscores as emphasis` pins the preset
  - `test/js/bun/md/md-react.test.ts`: `math` (with `display`) and `u` elements, default tags and component overrides
  - `test/js/bun/md/md-render-callback.test.ts`: math text reaches the `text` callback, `_` no longer reaches `emphasis`
  - `test/cli/run/markdown-entrypoint.test.ts`: `bun ./file.md` styles math and still renders `_x_` as italic
  - `bun test test/integration/bun-types/bun-types.test.ts` passes (and fails when the fixture's `display` assertion is changed)
  - The whole `test/js/bun/md/` directory passes on the debug build; no existing snapshot changes

### Background

- Inline parsing in this crate runs per slice (a paragraph, heading or table cell, or a link label inside one) in two phases: `collect_emphasis_delimiters` records every `*`, `_`, `~` run as an `EmphDelim`, `resolve_emphasis_delimiters` pairs them with the CommonMark algorithm, and a second walk over the text emits spans at the recorded positions. This PR adds `$` runs to that list; they are paired by their own rule but share the bookkeeping that stops emphasis and math from crossing each other.
- md4c is the C CommonMark parser this crate was ported from. Its `MD_FLAG_LATEXMATHSPANS` produces `MD_SPAN_LATEXMATH` / `MD_SPAN_LATEXMATH_DISPLAY` with the span's text as `MD_TEXT_LATEXMATH` (verbatim, newlines as spaces, like a code span); its `MD_FLAG_UNDERLINE` makes underscores produce `MD_SPAN_U`. The `SpanType` and `TextType` enums here mirror those, and the `<x-equation>` HTML is what md4c's md2html prints.
- `Options` (`src/md/root.rs`) is the user-facing option set; `Flags` (`src/md/types.rs`) is the subset the parser reads. An option that exists in `Options` but not in `Flags` is silently ignored, which is how `underline` was lost.

<details>
<summary>Flood timings on the debug (ASAN) build, n = 50k / 100k / 200k repetitions</summary>

```
openers then closers                            74 / 147 / 295 ms
mismatched pending opener                       37 /  77 / 145 ms
openers popped by emphasis                     409 / 832 / 1664 ms
alternating pairs                              158 / 322 / 646 ms
emphasis closers over pending math openers     133 / 268 / 526 ms
pending $$ openers and $ closers interleaved    76 / 155 / 309 ms
```

Inputs, in order: `"$a "*n + "a$ "*n`, `"$$a " + "a$ "*n`, `"*$a $a $a $a a* "*n`, `"$a$ "*n`, `"$a "*n + "a* "*n`, `"$$a a$ "*n`.

</details>

---

**[review]** gate passed · iteration 1 · 15 files touched

<details><summary>fails on main (without fix)</summary>

```console
ASAN without fix: BUILD FAILED (no junit output)
$ BUN_DEBUG_QUIET_LOGS=1 bun scripts/build.ts --profile=debug --quiet test "--reporter=junit" "--reporter-outfile=/tmp/mechgate.xml" test/cli/run/markdown-entrypoint.test.ts test/js/bun/md/md-edge-cases.test.ts test/js/bun/md/md-react.test.ts test/js/bun/md/md-render-callback.test.ts test/js/bun/md/md-spec.test.ts
ninja: Entering directory `/workspace/bun/build/debug'
[1/180] gen .bind.ts → GeneratedBindings.cpp
[2/180] gen JSBuffer.lut.h
Generating /workspace/bun/build/debug/codegen/JSBuffer.lut.h from /workspace/bun/src/jsc/bindings/JSBuffer.cpp
[3/180] gen cpp.rs (cppbind)
[4/180] gen BunProcess.lut.h
Generating /workspace/bun/build/debug/codegen/BunProcess.lut.h from /workspace/bun/src/jsc/bindings/BunProcess.cpp
[5/180] gen generated_host_exports.rs
generated_host_exports.rs: 92 exports (host=3, lazy=10, generic=79, rust=0); 240 extern-C blocks audited
[6/180] gen JS modules (bundle-modules)
Preprocess modules (12787ms)
Bundle modules (68ms)
Postprocesss modules (177ms)
Bundle Functions (952ms)
Generate Code (22ms)

[14.03s] Bundled "src/js" for development
  2819 kb
  198 internal modules
  13 native modules
  92 internal functions across 17 files
[6/180] ca
... (truncated)

release without fix: 1 skipped
bun test v1.4.0-canary.1 (9f0c2c74c)

test/cli/run/markdown-entrypoint.test.ts:
(pass) bun <file.md> > renders headings with underlines [4.92ms]
(pass) bun <file.md> > renders bold, italic, strikethrough, inline code [3.12ms]
(pass) bun <file.md> > renders math spans; underscores stay emphasis [2.71ms]
(pass) bun <file.md> > renders ordered, unordered, and task lists [2.27ms]
(pass) bun <file.md> > renders blockquotes and nested blockquotes [2.10ms]
(pass) bun <file.md> > renders horizontal rules [2.18ms]
(pass) bun <file.md> > renders fenced code block with JS syntax highlighting [2.01ms]
(pass) bun <file.md> > renders fenced code block without language [2.05ms]
(pass) bun <file.md> > renders link text with url fallback when no TTY [2.12ms]
(pass) bun <file.md> > emits OSC 8 escape for links when hyperlinks are enabled [20.14ms]
(pass) bun <file.md> > renders link with text + url pair fallback [4.57ms]
(pass) bun <file.md> > renders images as alt text with link [3.04ms]
(pass) bun <file.md> > renders wikilinks [3.31ms]
(pass) bun <file.md> > falls back to literal text when `[[` never closes [2.85ms]
(pass) bun <file.md> > renders simple table with alignment [2.19ms
... (truncated)
```

</details>

<details><summary>passes on PR (with fix)</summary>

```console
ASAN with fix: all passed
$ BUN_DEBUG_QUIET_LOGS=1 bun scripts/build.ts --profile=debug --quiet test "--reporter=junit" "--reporter-outfile=/tmp/mechgate.xml" test/cli/run/markdown-entrypoint.test.ts test/js/bun/md/md-edge-cases.test.ts test/js/bun/md/md-react.test.ts test/js/bun/md/md-render-callback.test.ts test/js/bun/md/md-spec.test.ts
bun test v1.4.0 (6e906e468)

test/cli/run/markdown-entrypoint.test.ts:
(pass) bun <file.md> > renders headings with underlines [171.73ms]
(pass) bun <file.md> > renders bold, italic, strikethrough, inline code [115.50ms]
(pass) bun <file.md> > renders math spans; underscores stay emphasis [119.48ms]
(pass) bun <file.md> > renders ordered, unordered, and task lists [116.59ms]
(pass) bun <file.md> > renders blockquotes and nested blockquotes [117.08ms]
(pass) bun <file.md> > renders horizontal rules [111.13ms]
(pass) bun <file.md> > renders fenced code block with JS syntax highlighting [116.53ms]
(pass) bun <file.md> > renders fenced code block without language [125.89ms]
(pass) bun <file.md> > renders link text with url fallback when no TTY [136.40ms]
(pass) bun <file.md> > emits OSC 8 escape for links when hyperlinks are enabled [1107.25ms]
(pass) bun <fil
... (truncated)

release with fix: 1 skipped
$ bun scripts/build.ts --profile=release
[configured] bun-profile → bun (stripped) in 699ms (unchanged)
ninja: Entering directory `/workspace/bun/build/release'
[1/139] gen .bind.ts → GeneratedBindings.cpp
[2/139] gen JSBuffer.lut.h
Generating /workspace/bun/build/release/codegen/JSBuffer.lut.h from /workspace/bun/src/jsc/bindings/JSBuffer.cpp
[3/139] gen generated_host_exports.rs
generated_host_exports.rs: 92 exports (host=3, lazy=10, generic=79, rust=0); 240 extern-C blocks audited
[4/139] gen BunProcess.lut.h
Generating /workspace/bun/build/release/codegen/BunProcess.lut.h from /workspace/bun/src/jsc/bindings/BunProcess.cpp
[5/139] gen cpp.rs (cppbind)
[6/139] gen JS modules (bundle-modules)
Preprocess modules (8484ms)
Bundle modules (145ms)
Postprocesss modules (344ms)
Bundle Functions (867ms)
Generate Code (29ms)

[9.89s] Bundled "src/js" for production
  2628 kb
  198 internal modules
  13 native modules
  92 internal functions across 17 files
[6/138] cargo bun_bin → libbun_rust.a (--target x86_64-unknown-linux-gnu)

  nightly-2026-07-20-x86_64-unknown-linux-gnu unchanged - rustc 1.99.0-nightly (9f36de775 2026-07-19)

[1m[92m   Compiling[0m bun_core v
... (truncated)
```

</details>

<details><summary>diff hotspot</summary>

```
docs/runtime/markdown.mdx                      |   6 +-
 packages/bun-types/bun.d.ts                    |  24 ++-
 src/md/ansi_renderer.rs                        |  22 ++-
 src/md/helpers.rs                              |   4 +
 src/md/html_renderer.rs                        |   4 +-
 src/md/inlines.rs                              | 117 +++++++++++++-
 src/md/parser.rs                               |   2 +-
 src/md/root.rs                                 |   4 +-
 src/md/types.rs                                |   1 +
 test/cli/run/markdown-entrypoint.test.ts       |   4 +
 test/integration/bun-types/fixture/markdown.ts |  27 ++++
 test/js/bun/md/md-edge-cases.test.ts           | 208 ++++++++++++++++++++++++-
 test/js/bun/md/md-react.test.ts                |  76 +++++++++
 test/js/bun/md/md-render-callback.test.ts      |  32 ++++
 test/js/bun/md/md-spec.test.ts                 |   2 +
 15 files changed, 510 insertions(+), 23 deletions(-)
```

</details>

**gate history** · 3 passed · 0 rejected · iteration 1

<details><summary>evidence per changed file</summary>

```
file                                            reads  edits  tests
docs/runtime/markdown.mdx                           2      2      0
packages/bun-types/bun.d.ts                         5      5      0
src/md/ansi_renderer.rs                             4      2      0
src/md/helpers.rs                                   1      1      0
src/md/html_renderer.rs                             1      1      0
src/md/inlines.rs                                  14     25      0
src/md/parser.rs                                    2      2      0
src/md/root.rs                                      2      4      0
src/md/types.rs                                     1      2      0
test/cli/run/markdown-entrypoint.test.ts            2      3      0
test/integration/bun-types/fixture/markdown.ts      0      2      0
test/js/bun/md/md-edge-cases.test.ts                6     11      0
test/js/bun/md/md-react.test.ts                     3      5      0
test/js/bun/md/md-render-callback.test.ts           1      1      0
test/js/bun/md/md-spec.test.ts                      1      1      0
```

</details>

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 9:57 PM PT - Aug 17th, 2026</sup></div>

@robobun, your commit a2888463fa8ba4566ed6d9de428c00ca546c90c9 is building: [`#100391`](https://buildkite.com/bun/bun/builds/100391)

**Comment by robobun** [comment]
Status: fix is up in this PR, self-review in progress.

Reproduced on bun 1.4.0 and on main at 9a895c23:

```
$ bun -e 'console.log(Bun.markdown.html("$a+b$\n", { latexMath: true }))'
<p>$a+b$</p>
$ bun -e 'console.log(Bun.markdown.html("__x__\n", { underline: true }))'
<p><strong>x</strong></p>
```

Fail-before / pass-after was checked per file: `md-spec.test.ts` (6 examples from the two checked-in md4c spec files), `md-edge-cases.test.ts` (32 cases), `md-react.test.ts` (10), `md-render-callback.test.ts` (2) and `test/cli/run/markdown-entrypoint.test.ts` (1) fail on the released binary and pass with this branch's debug build; the rest of `test/js/bun/md/` is unchanged and green.

Related findings filed separately: #39491 (`hardSoftBreaks` / `collapseWhitespace` are no-ops too) and #39492 (emphasis across wiki link targets, pre-existing).

File: src/md/inlines.rs [hunk 1]
```diff
@@ -37,6 +37,9 @@ pub struct EmphDelim {
     pub(crate) close_sizes: [u8; MAX_EMPH_MATCHES],
     pub(crate) close_num: u8, // number of close matches
     pub(crate) active: bool,  // false if deactivated between matched pairs
+    // `$` openers only: position of the closing run this opener was matched
+    // with. The span's content is `content[pos + count..close_pos]`.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/inlines.rs]: Trimmed to a single line in 43cb295 (the comment check reports 0 multi-line groups in the current diff).

File: src/md/inlines.rs [hunk 2]
```diff
@@ -331,6 +335,40 @@ impl Parser<'_> {
                     continue;
                 }
 
+                // Math span (`$...$` or `$$...$$`), paired up in
+                // resolve_math_delim. The whole span is emitted here and its
+                // content is verbatim, so the walk skips to the closing run. A
+                // run that did not open a span is text; that includes a closer
+                // whose opener was consumed by a construct emitted before it
+                // (a wiki link whose target contains a `$`).
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/inlines.rs]: Trimmed to a single line in 43cb295 (the comment check reports 0 multi-line groups in the current diff).

File: src/md/inlines.rs [hunk 3]
```diff
@@ -611,12 +655,28 @@ impl Parser<'_> {
         self.renderer.text(text_type, content)
     }
 
+    /// With the underline extension, `_` runs still pair up by the CommonMark
+    /// emphasis rules, but every matched underscore becomes one `U` span
+    /// instead of an em (1) or strong (2) span, as in md4c's MD_FLAG_UNDERLINE.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/inlines.rs]: Trimmed to a single line in 43cb295 (the comment check reports 0 multi-line groups in the current diff).

File: src/md/inlines.rs [hunk 4]
```diff
@@ -756,6 +826,39 @@ impl Parser<'_> {
                 });
                 continue;
             }
+            // Math span delimiter (1 or 2 dollars only; a longer run is text).
+            // md4c's rule: a run can open only after whitespace or punctuation
+            // and close only before them, so `$5 and $10` stays text.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/inlines.rs]: Trimmed to a single line in 43cb295 (the comment check reports 0 multi-line groups in the current diff).

File: src/md/inlines.rs [hunk 5]
```diff
@@ -919,6 +1032,43 @@ impl Parser<'_> {
         }
     }
 
+    /// Pair `$` runs the way md4c's `md_analyze_dollar` does: a run closes the
+    /// most recent pending opener only if both runs have the same length
+    /// (`$` vs `$$`); otherwise it becomes a pending opener itself, if it may
+    /// open. A match discards every pending opener, so math spans never nest.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/inlines.rs]: Trimmed to a single line in 43cb295 (the comment check reports 0 multi-line groups in the current diff).

File: src/md/inlines.rs [hunk 6]
```diff
@@ -919,6 +1032,43 @@ impl Parser<'_> {
...
+                let opener = &mut self.emph_delims[oi];
+                debug_assert!(opener.active && opener.remaining > 0);
+                if opener.count == run.count {
+                    opener.remaining = 0;
+                    opener.open_count = run.count;
+                    opener.close_pos = run.pos;
+                    self.emph_delims[idx].remaining = 0;
+                    self.emph_delims[idx].close_count = run.count;
+                    // The span's content is verbatim, so nothing inside it may
+                    // still pair up with a delimiter outside of it.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/inlines.rs]: Trimmed to a single line in 43cb295 (the comment check reports 0 multi-line groups in the current diff).

File: src/md/parser.rs [hunk 7]
```diff
@@ -382,8 +382,8 @@ impl<'a> Parser<'a> {
     //   emit_text, emit_emph_open_tags, emit_emph_close_tags,
     //   find_code_span_end, normalize_code_span_content, is_left_flanking,
     //   is_right_flanking, can_open_emphasis, can_close_emphasis,
-    //   collect_emphasis_delimiters, resolve_emphasis_delimiters, find_entity,
-    //   find_html_tag
+    //   collect_emphasis_delimiters, resolve_emphasis_delimiters,
+    //   resolve_math_delim, find_entity, find_html_tag
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/parser.rs]: Trimmed to a single line in 43cb295 (the comment check reports 0 multi-line groups in the current diff).

**Comment by robobun** [comment]
Addressed in 43cb295:

- The underline finding from the review (the `U` span count came from `open_sizes` / `close_sizes`, which hold at most `MAX_EMPH_MATCHES` = 6 entries, so 13+ underscores per side emitted 12 spans and a run closed by seven closers came out unbalanced) is valid. Underline is now emitted in the walk from the run's `open_count` / `close_count`, one `U` per consumed underscore, which is also md4c's definition; `emit_emph_open_tags` / `emit_emph_close_tags` are back to their main versions. Two regression cases were added in md-edge-cases.test.ts (14 underscores per side, and one 14-underscore run consumed by seven `__` closers). The same cap still affects em/strong output on main independently of this PR; filed as #39496.
- The seven comment-check threads: each flagged comment is one line now (the check reports 0 multi-line groups on the current diff), threads replied to and resolved.
- PR description updated to match; #39494 (another PR) picks up #39492, and its skip in the collection phase will cover `$` runs as well.

**Comment by robobun** [comment]
For whoever takes the human look at the two spots called out above:

- `math_openers.clear()` on a match is md4c's `DOLLAR_OPENERS.top = -1` in `md_analyze_dollar` (it is what makes `$$foo $bar$ baz$$` render the inner span and leave the outer `$$` as text, spec-latex-math.txt example 2). Popping only the matched opener would let the outer `$$` pair up afterwards and swallow the inner span.
- `prev_candidate[idx] = oi` mirrors what the emphasis path does after its own matches: every entry strictly between the `$` opener and closer has just been deactivated, so a later emphasis closer walking the candidate chain can hop over the span in one step. Correctness does not depend on it (a walk through the dead entries unlinks them one by one); it only keeps that walk amortized O(1). The math pairing itself never consults the chain, only `math_openers`.

CI is green on 43cb295 (Buildkite #100415).

File: test/cli/run/markdown-entrypoint.test.ts [hunk 8]
```diff
@@ -41,6 +41,10 @@ describe("bun <file.md>", () => {
     expect(await runMd("**bold** *italic* ~~strike~~ `code` regular\n")).toMatchSnapshot();
   });
 
+  test("renders math spans; underscores stay emphasis", async () => {
+    expect(await runMd("_em_ and $x$\n")).toBe("\x1b[3mem\x1b[23m and \x1b[35m$x$\x1b[39m\x1b[0m\n");
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/cli/run/markdown-entrypoint.test.ts:45]: Leaving this one as is. The test goes through the file's shared `runMd()` helper, which drains stdout, stderr and the exit code concurrently and asserts the exit code in `afterEach`; it deliberately does not assert stderr (comment at test/cli/run/markdown-entrypoint.test.ts:19, about ASAN lanes), and the other ~30 tests in the file share that contract. Changing the helper's contract for the whole file is a separate cleanup, not something this PR should do for one of its tests. For what it is worth, stderr is empty for this input on the local ASAN debug build, so that cleanup looks feasible if someone wants it.

**Comment by robobun** [comment]
Follow-up to the full re-review after the rebase:

- The ANSI staging buffer now grows through `try_extend`, so an allocation failure while joining a multi-line code/math span lands on `OutputBuffer::oom` like every other staging buffer in the renderer and `render_to_ansi` reports `OutOfMemory` (9f0c2c7).
- The stderr suggestion on the `bun ./file.md` test was withdrawn in-thread: the test uses the file's shared `runMd()` helper, which intentionally does not assert stderr.

The rebase itself (ac01f33) had one conflict: #39795 removed the `Flags::DEFAULTS` block that the first commit had added a line to; main's deletion was kept, and only the new `Flags.underline` field remains in types.rs. No behavior changes since 6573a6b apart from the OOM path above; the md and entrypoint suites pass on the rebased debug build.
